### PR TITLE
Shared kernelconnection wrapper

### PR DESCRIPTION
**Pull request by DonJayamanne**
Part of #9503

### CODE REVIEW

File: src/kernels/debugger/kernelDebugAdapterBase.ts [hunk 1]
```diff
@@ -89,18 +90,9 @@ export abstract class KernelDebugAdapterBase implements DebugAdapter, IKernelDeb
             this.debugCell = notebookDocument.cellAt(configuration.__cellIndex!);
         }
 
+        this.jupyterSession.kernel?.iopubMessage.connect(this.onIOPubMessage, this);
         this.disposables.push(
-            this.jupyterSession.onIOPubMessage(async (msg: KernelMessage.IIOPubMessage) => {
```
Comment (DonJayamanne) [inline src/kernels/debugger/kernelDebugAdapterBase.ts:93]: This is what I'd like to fix in this PR.
We need a way to hook up to the iopub and anymessage events in the underlyiing Jupyter kernel connection.
This property `onIOPubMessage` is a custom event we wire up. If I were to introduce anyMessage then i'd have to handle kernel restarts etc... As we expose the underlying kernel everywhere it makes sense to just hook to those events.

Hence created a kernel wrapper so that we have a single instance, of the Kernel.KernleConnection on which we can monitor the events.

File: src/kernels/common/baseJupyterSession.ts [hunk 2]
```diff
@@ -83,28 +91,30 @@ export abstract class BaseJupyterSession implements IJupyterSession {
...
+        return this._wrappedKernel;
     }
+
     public get kernelSocket(): Observable<KernelSocketInformation | undefined> {
         return this._kernelSocket;
     }
     public get onSessionStatusChanged(): Event<KernelMessage.Status> {
         return this.onStatusChangedEvent.event;
     }
-    public get onIOPubMessage(): Event<KernelMessage.IIOPubMessage> {
```
Comment (DonJayamanne) [inline src/kernels/common/baseJupyterSession.ts:95]: We need a way to hook up to the `iopub` and `anymessage` events in the underlyiing Jupyter kernel connection.
This property `onIOPubMessage` is a custom event we wire up. If I were to introduce `anyMessage` event then i'd have to handle kernel restarts all over again, tomorrow if we wanted to handle other events, then all over gain etc... 

Also, as we expose the underlying jupyter kernel (Kernel.IKenrelConnection) everywhere it makes sense to just hook to those events.

Hence created a kernel wrapper so that we have a single instance, of the Kernel.KernleConnection on which we can monitor the events (basically moved onIOPubMessage & similar events into the kernel.IKernelConnection and have that keep track of all the event handlers even when kernels restart)

File: src/kernels/debugger/kernelDebugAdapterBase.ts [hunk 3]
```diff
@@ -89,18 +90,9 @@ export abstract class KernelDebugAdapterBase implements DebugAdapter, IKernelDeb
...
-                const anyMsg = msg as any;
-                traceInfoIfCI(`Debug IO Pub message: ${JSON.stringify(msg)}`);
-                if (anyMsg.header.msg_type === 'debug_event') {
-                    this.trace('event', JSON.stringify(msg));
-                    if (!(await this.delegate?.willSendEvent(anyMsg))) {
-                        this.sendMessage.fire(msg.content);
-                    }
-                }
-            })
+            new Disposable(() => this.jupyterSession.kernel?.iopubMessage.disconnect(this.onIOPubMessage, this))
```
Comment (DonJayamanne) [inline src/kernels/debugger/kernelDebugAdapterBase.ts:95]: Now we can hook up to the underlying Jupyter API events.

File: src/platform/api/kernelConnectionWrapper.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import type { Kernel } from '@jupyterlab/services';
+import { IDisposable } from '../common/types';
+import { noop } from '../common/utils/misc';
+import { BaseKernelConnectionWrapper } from '../../kernels/jupyter/baseKernelConnectionWrapper';
+import { IKernel } from '../../kernels/types';
+
+export class KernelConnectionWrapper extends BaseKernelConnectionWrapper {
```
Comment (DonJayamanne) [inline src/platform/api/kernelConnectionWrapper.ts:10]: wrapper around IKernel for 3rd party API.

File: src/kernels/raw/session/rawSocket.node.ts [hunk 5]
```diff
@@ -100,7 +101,7 @@ export class RawSocket implements IWebSocketLike, IKernelSocket, IDisposable {
         // If from ipywidgets, this will be serialized already, so turn it back into a message so
         // we can add the special hash to it.
         const message = this.deserialize(data);
-
+        this.onAnyMessage(message as any);
```
Comment (DonJayamanne) [inline src/kernels/raw/session/rawSocket.node.ts:104]: When we get messages from the UI kenel, we send that directly to the kernel bypassing the Jupyter lab API
As a result we don't invoke the event `anymessage` (which happens in jupyter code when messages are sent to the kernel)

For https://github.com/microsoft/vscode-jupyter/issues/9503, we need the ability to monitor all messagesa & anymessage event handler will pick up some of these messages which only works in Jupyter & not in raw kernels.

Basically this fixes an issue.

File: src/kernels/raw/session/rawJupyterSession.node.ts [hunk 6]
```diff
@@ -51,9 +51,6 @@ export class RawJupyterSession extends BaseJupyterSession {
         }
         return super.status;
     }
-    public get kernelId(): string {
-        return this.session?.kernel?.id || '';
```
Comment (DonJayamanne) [inline src/kernels/raw/session/rawJupyterSession.node.ts:55]: Duplicate code in both child classes, moved to base class

File: src/kernels/jupyter/session/jupyterSession.ts [hunk 7]
```diff
@@ -76,15 +69,6 @@ export class JupyterSession extends BaseJupyterSession implements IJupyterServer
         // Wait for idle on this session
         return this.waitForIdleOnSession(this.session, timeout);
     }
-
-    public override get kernel(): Kernel.IKernelConnection | undefined {
-        return this.session?.kernel || undefined;
```
Comment (DonJayamanne) [inline src/kernels/jupyter/session/jupyterSession.ts:81]: Duplicate code in both child classes, moved to base class

File: src/kernels/jupyter/baseKernelConnectionWrapper.ts [hunk 8]
```diff
@@ -74,27 +57,8 @@ export class KernelConnectionWrapper implements Kernel.IKernelConnection {
         return this.possibleKernelConnection;
     }
 
-    constructor(readonly kernel: IKernel, disposables: IDisposable[]) {
-        const emiStatusChangeEvents = () => {
-            this.statusChanged.emit(kernel.status);
-            if (kernel.status === 'dead' && !kernel.disposed && !kernel.disposing) {
-                this.connectionStatusChanged.emit('disconnected');
```
Comment (rchiodo) [inline src/kernels/jupyter/baseKernelConnectionWrapper.ts:81]: Where did this logic go?

Comment (rchiodo) [reply src/kernels/jupyter/baseKernelConnectionWrapper.ts:81]: Assuming we still need to track connection status.

Comment (rchiodo) [reply src/kernels/jupyter/baseKernelConnectionWrapper.ts:81]: Oh never mind. This moved to the new class.

**Review by amunger — APPROVED** [review]

File: src/kernels/raw/session/rawSocket.node.ts (hunk 5 above)
Comment (rchiodo) [reply src/kernels/raw/session/rawSocket.node.ts:104]: Can you put this statement as a comment? I'm probably going to wonder why we're calling this.onAnyMessage here later on.
